File: source/php/perm.php
<?php
//ini_set('xdebug.max_nesting_level', 100);
ini_set('xdebug.max_nesting_level', -1);
// ini_set('memory_limit', '256M');
ini_set('memory_limit', '4096M');


if ($argc < 3) {
    die("Usage: php perm.php  <R|I> <n>\nR=Recursive, I=Iterative\nn=size of input array)");
}

$mode = strtoupper($argv[1]);
if ($mode !== "I"&& $mode !== "R") {
    die("invalid value, must be <R> or <I>");
}

$n = $argv[2];
if ($n < 0) {
    $n = 0;
}

$array = array();
for ($i = 0; $i < $n; $i++) {
    $array[] = $i;
}

// echo "Input: [" . implode(", ", $array) . "]\n";

if ($mode === "I") {
    $perm = 'permRec';
} elseif ($mode === 'R') {
    $perm = 'permIt';
} else {
    die();
}

try {
    $start = hrtime(true);
    $result = $perm($array);
    $end = hrtime(true);
    $elapsedTime = ($end - $start) / 1e9;
    echo "$mode: " . count($result) . " permutations in $elapsedTime seconds\n";
} catch (Exception $e) {
    echo $e->getMessage();
}


function permRec($array)
{
    if (count($array) <= 1) {
        return [$array];
    }
    $result = [];
    foreach ($array as $key => $element) {
        foreach (permRec(array_diff_key($array, [$key => $element])) as $perm) {
            $result[] = array_merge([$element], $perm);
        }
    }
    return $result;
}

function permIt($array)
{
    $result = [];
    $count = count($array);
    $c = array_fill(0, $count, 0);
    $result[] = $array;

    $i = 0;
    while ($i < $count) {
        if ($c[$i] < $i) {
            if ($i % 2 == 0) {
                $temp = $array[0];
                $array[0] = $array[$i];
                $array[$i] = $temp;
            } else {
                $temp = $array[$c[$i]];
                $array[$c[$i]] = $array[$i];
                $array[$i] = $temp;
            }

            $result[] = $array;
            $c[$i]++;
            $i = 0;
        } else {
            $c[$i] = 0;
            $i++;
        }
    }
    return $result;
}

File: source/rust/perm.rs
use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() < 3 {
        eprintln!("Usage: perm <R|I> <n>\nR=Recursive, I=Iterative\nn=size of input array)");
        std::process::exit(1);
    }

    let mode = args[1].to_uppercase();
    if mode != "I" && mode != "R" {
        eprintln!("Invalid value, must be <R> or <I>");
        std::process::exit(1);
    }

    let n: usize = args[2].parse().unwrap_or(0);

    let array: Vec<usize> = (0..n).collect();

    let perm: fn(Vec<usize>) -> Vec<Vec<usize>> = if mode == "I" {
        perm_it
    } else {
        perm_rec
    };

    let start = Instant::now();
    let result = perm(array.clone());
    let elapsed_time = start.elapsed().as_secs_f64();
    println!("{}: {} permutations in {} seconds", mode, result.len(), elapsed_time);
}

fn perm_rec(array: Vec<usize>) -> Vec<Vec<usize>> {
    if array.len() <= 1 {
        return vec![array];
    }
    let mut result = Vec::new();
    for (key, &element) in array.iter().enumerate() {
        for perm in perm_rec(array.iter().enumerate().filter_map(|(i, &e)| if i != key { Some(e) } else { None }).collect()) {
            let mut new_perm = vec![element];
            new_perm.extend(perm);
            result.push(new_perm);
        }
    }
    result
}

fn perm_it(mut array: Vec<usize>) -> Vec<Vec<usize>> {
    let mut result = Vec::new();
    let count = array.len();
    let mut c: Vec<usize> = vec![0; count];

    result.push(array.clone());

    let mut i = 0;
    while i < count {
        if c[i] < i {
            if i % 2 == 0 {
                array.swap(0, i);
            } else {
                array.swap(c[i], i);
            }

            result.push(array.clone());
            c[i] += 1;
            i = 0;
        } else {
            c[i] = 0;
            i += 1;
        }
    }
    result
}